src/space_usage/mod.rs

```diff
@@ -0,0 +1,283 @@
+/*!
+Representations for the space usage of various parts of a Tantivy index.
+
+This can be used programmatically, and will also be exposed in a human readable fashion in
+tantivy-cli.
+
+One important caveat for all of this functionality is that none of it currently takes storage-level
+details into consideration. For example, if your file system block size is 4096 bytes, we can
+under-count actual resultant space usage by up to 4095 bytes per file.
+*/
+
+use schema::Field;
+use std::collections::HashMap;
+use std::ops::{Add, AddAssign};
+use serde::Serialize;
+use serde::Serializer;
+use serde::Deserialize;
+use serde::Deserializer;
+
+/// Indicates space usage in bytes
+#[derive(Clone, Copy, Debug)]
+pub struct ByteCount(pub usize);
+
+impl Serialize for ByteCount {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: Serializer {
+        self.0.serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for ByteCount {
+    fn deserialize<D>(deserializer: D) -> Result<ByteCount, D::Error> where D: Deserializer<'de> {
+        Ok(ByteCount(usize::deserialize(deserializer)?))
+    }
+}
+
+impl Add for ByteCount {
+    type Output = ByteCount;
+    fn add(self, rhs: ByteCount) -> ByteCount {
+        ByteCount(self.0 + rhs.0)
+    }
+}
+
+impl AddAssign for ByteCount {
+    fn add_assign(&mut self, rhs: ByteCount) {
+        self.0 += rhs.0;
+    }
+}
+
+/// Represents combined space usage of an entire searcher and its component segments.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SearcherSpaceUsage {
+    segments: Vec<SegmentSpaceUsage>,
+    total: ByteCount,
+}
+
+impl SearcherSpaceUsage {
+    pub(crate) fn new() -> SearcherSpaceUsage {
+        SearcherSpaceUsage {
+            segments: Vec::new(),
+            total: ByteCount(0),
+        }
+    }
+
+    /// Add a segment, to `self`.
+    /// Performs no deduplication or other intelligence.
+    pub(crate) fn add_segment(&mut self, segment: SegmentSpaceUsage) {
+        self.total += segment.total();
+        self.segments.push(segment);
+    }
+
+    /// Per segment space usage
+    pub fn segments(&self) -> &[SegmentSpaceUsage] {
+        &self.segments[..]
+    }
+
+    /// Returns total byte usage of this searcher, including all large subcomponents.
+    /// Does not account for smaller things like `meta.json`.
+    pub fn total(&self) -> ByteCount {
+        self.total
+    }
+}
+
+/// Represents combined space usage for all of the large components comprising a segment.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SegmentSpaceUsage {
+    num_docs: u32,
+
+    termdict: PerFieldSpaceUsage,
```

Comment:
Can we use a HashMap<SegmentComponent, PerFieldSpaceUsage> instead?

Reply:
This is possible, but gives up some static safety of having actually filled in all the values.

Reply:
That's a matter of perspective. If I add a new component today, nothing enforces fixing the `SegmentSpaceUsage` structure.

You can get the best of two world (kind of) by populating it with a switch case.

```rust
fn compute_segment_space_usage(&self, component: SegmentComponent) {
    match component {
      // forces you to handle all components.
   }
}
```

Populating the HashMap can then be done by iterating over the SegmentComponent values.

```rust
for component in SegmentComponent::iterator() {
    let component_space_usage = self.compute_segment_space_usage(component);
    map.insert(component, component_space_usage )
}
```